plugins/woocommerce-blocks/assets/js/events/event-emitter.ts

```
@@ -0,0 +1,121 @@
+/**
+ * External dependencies
+ */
+import {
+	isErrorResponse,
+	isFailResponse,
+	isObserverResponse,
+	type ObserverResponse,
+} from '@woocommerce/types';
+
+export interface EventEmitter {
+	emit: (
+		eventName: string,
+		data?: unknown
+	) => Promise< ObserverResponse[] >;
+	emitWithAbort: (
+		eventName: string,
+		data?: unknown
+	) => Promise< ObserverResponse[] >;
+	subscribe: (
+		listener: EventListener,
+		priority: number,
+		eventName: string
+	) => VoidFunction;
+}
+
+export type EventListener = (
+	data: unknown
+) => Promise< ObserverResponse | true > | ObserverResponse | true;
+
+export interface EventListenerWithPriority {
+	listener: EventListener;
+	priority: number;
+}
+/**
+ * Create an event emitter.
+ *
+ * @return The event emitter.
+ */
+export function createEmitter(): EventEmitter {
```

The inspiration for this was taken from https://github.com/WordPress/gutenberg/blob/trunk/packages/data/src/utils/emitter.ts#L14 which drive's `@wordpress/data`'s `subscribe` however we currently don't need pause/resume functionality and we need to add a way to emit events and fail fast on error (`emitWithAbort`)

The way this is designed to be used is to create a new event emitter for each distinct set of events you want to handle, e.g. this PR adds a `checkoutEventsEmitter`, we could also have a `paymentEventsEmitter` and a `cartEventsEmitter` eventually.

The Gutenberg implementation also stores the listeners in a `Set` which is not something we can do as we need our listeners to have priority. Therefore, we need to sort them. Given a `Set` is unordered, this sort would need to happen on every emit, using an array instead allows us to sort once on insert, improving performance.